### PR TITLE
[5.0] Static Dispatch for the lock of `Atomic`.

### PR DESCRIPTION
**Pull request by andersio**
Justification: #3076

This branch uses type erasure, and adds an internal `RecursiveAtomic` for `MutableProperty`. The downside is that `_AtomicBase` would be leaked to the public API.

An alternative would be a parameterised `Atomic` with the lock type. e.g. `Atomic<Value, PosixThreadMutex>`.

### CODE REVIEW

**Comment by mdiep** [comment]
I'm pretty 👎🏻 on leaking details like that. 😕 

If it came down to it, I'd rather duplicate the code in `Atomic` or parameterize `Atomic` like you suggested. It's too bad that Foundation doesn't expose something like `PosixThreadMutex`.

**Comment by andersio** [comment]
I have opted for duplication in take 2, since parametrising `Atomic` would lose the convenient shorthand e.g. `let interrupted = Atomic(false)`. We got some shared impl. in AtomicProtocol though.

But due to overload ambiguity, the `completion` parameter of `modify` is removed, and `RecursiveAtomic` accepts a didSet observer in its initializer instead.

File: ReactiveCocoa/Swift/Property.swift [hunk 1]
```diff
@@ -580,14 +580,14 @@ public final class MutableProperty<Value>: MutablePropertyProtocol {
...
 		/// `value`. Note that recursive sets will still deadlock because the
 		/// underlying producer prevents sending recursive events.
-		let lock = RecursiveLock()
-		lock.name = "org.reactivecocoa.ReactiveCocoa.MutableProperty"
-		_atomic = Atomic(initialValue, mutex: lock)
-
-		(signal, observer) = Signal.pipe()
+		_atomic = RecursiveAtomic(initialValue,
+															name: "org.reactivecocoa.ReactiveCocoa.MutableProperty",
+															didSet: observer.sendNext)
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Property.swift]: [Indentation here is wrong](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/dc104c86b59533a5de2c75813d65bf96b24c516d/ReactiveCocoa/Swift/Property.swift#L589).

File: ReactiveCocoa/Swift/Atomic.swift [hunk 2]
```diff
@@ -33,77 +33,79 @@ final class PosixThreadMutex: Locking {
...
+	/// Initialize the variable with the given initial value.
+	/// 
+	/// - parameters:
+	///   - value: Initial value for `self`.
+	///   - name: An optional name used to create the recursive lock.
+	internal init(_ value: Value, name: StaticString? = nil, didSet action: ((Value) -> Void)? = nil) {
+		_value = value
+
+		lock = RecursiveLock()
+		lock.name = name.map { String($0) }
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Atomic.swift]: This can be `name.map(String.init)`, no?

Comment (andersio) [inline ReactiveCocoa/Swift/Atomic.swift]: It cannot. Overloading doesn't seem working.

Comment (andersio) [inline ReactiveCocoa/Swift/Atomic.swift]: We need to specify which variant though. So it would be `String.init(_:)`.

**Comment by andersio** [comment]
The changes in the property names are bikeshedding. But there seems having no point to underscore the property except for a name clash with public API. (Note: AFAIK we previously use them only for `_value`.) The stdlib also uses this as a convention for the leaked implementation details, e.g. `Array._buffer`, which aren't the case of these changes.

**Pull request merged**